File: frontend/src/api/student/ai-advanced.ts
import { apiClient } from '../apiClient';

export interface AIRecommendation {
  id: number;
  user_id: number;
  recommendation_type: string;
  title: string;
  description?: string;
  content_id?: number;
  quiz_id?: number;
  learning_path_id?: number;
  confidence_score: number;
  reason?: string;
  is_accepted: boolean;
  is_dismissed: boolean;
  created_at: string;
}

export interface AITutoringSession {
  id: number;
  user_id: number;
  subject?: string;
  topic?: string;
  session_type: string;
  start_time: string;
  end_time?: string;
  duration?: number;
  status: string;
  notes?: string;
}

export interface AITutoringInteraction {
  id: number;
  session_id: number;
  user_message: string;
  ai_response: string;
  interaction_type: string;
  timestamp: string;
  user_satisfaction?: number;
}

export interface DifficultyDetection {
  id: number;
  user_id: number;
  subject: string;
  topic: string;
  difficulty_level: string;
  confidence_score: number;
  evidence?: any;
  detected_at: string;
  is_resolved: boolean;
  resolution_notes?: string;
}

export interface LearningAnalytics {
  id: number;
  user_id: number;
  analytics_type: string;
  data: any;
  insights?: string;
  recommendations?: any;
  created_at: string;
}

export interface AdaptiveContent {
  id: number;
  user_id: number;
  content_id: number;
  adaptation_type: string;
  original_content: any;
  adapted_content: any;
  adaptation_reason: string;
  created_at: string;
}

export interface CreateAIRecommendationData {
  recommendation_type: string;
  title: string;
  description?: string;
  content_id?: number;
  quiz_id?: number;
  learning_path_id?: number;
  confidence_score?: number;
  reason?: string;
}

export interface CreateAITutoringSessionData {
  subject?: string;
  topic?: string;
  session_type?: string;
}

export interface CreateAITutoringInteractionData {
  user_message: string;
  interaction_type?: string;
}

export interface CreateDifficultyDetectionData {
  subject: string;
  topic: string;
  difficulty_level: string;
  confidence_score?: number;
  evidence?: any;
}

class AIAdvancedAPI {
  // Recommandations IA
  async getAIRecommendations(recommendation_type?: string, limit?: number): Promise<AIRecommendation[]> {
    const params = new URLSearchParams();
    if (recommendation_type) params.append('recommendation_type', recommendation_type);
    if (limit) params.append('limit', limit.toString());
    
    return this.request(`/api/v1/ai-advanced/recommendations?${params.toString()}`);
  }

  async acceptAIRecommendation(recommendationId: number): Promise<void> {
    return this.request(`/api/v1/ai-advanced/recommendations/${recommendationId}/accept`, {
      method: 'POST',
    });
  }

  async dismissAIRecommendation(recommendationId: number): Promise<void> {
    return this.request(`/api/v1/ai-advanced/recommendations/${recommendationId}/dismiss`, {
      method: 'POST',
    });
  }

  // Sessions de tutorat IA
  async getAITutoringSessions(status?: string): Promise<AITutoringSession[]> {
    const params = new URLSearchParams();
    if (status) params.append('status', status);
    
    return this.request(`/api/v1/ai-advanced/tutoring/sessions?${params.toString()}`);
  }

  async createAITutoringSession(data: CreateAITutoringSessionData): Promise<AITutoringSession> {
    return this.request('/api/v1/ai-advanced/tutoring/sessions', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  async getAITutoringInteractions(sessionId: number): Promise<AITutoringInteraction[]> {
    return this.request(`/api/v1/ai-advanced/tutoring/sessions/${sessionId}/interactions`);
  }

  async createAITutoringInteraction(sessionId: number, data: CreateAITutoringInteractionData): Promise<AITutoringInteraction> {
    return this.request(`/api/v1/ai-advanced/tutoring/sessions/${sessionId}/interactions`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  // Détection de difficultés
  async getDifficultyDetections(subject?: string, is_resolved?: boolean): Promise<DifficultyDetection[]> {
    const params = new URLSearchParams();
    if (subject) params.append('subject', subject);
    if (is_resolved !== undefined) params.append('is_resolved', is_resolved.toString());
    
    return this.request(`/api/v1/ai-advanced/difficulty-detection?${params.toString()}`);
  }

  async createDifficultyDetection(data: CreateDifficultyDetectionData): Promise<DifficultyDetection> {
    return this.request('/api/v1/ai-advanced/difficulty-detection', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  async resolveDifficultyDetection(detectionId: number, resolution_notes?: string): Promise<void> {
    const body = resolution_notes ? JSON.stringify({ resolution_notes }) : undefined;
    return this.request(`/api/v1/ai-advanced/difficulty-detection/${detectionId}/resolve`, {
      method: 'PUT',
      body,
    });
  }

  // Analytics d'apprentissage
  async getPerformanceAnalytics(subject?: string, period?: string): Promise<any> {
    const params = new URLSearchParams();
    if (subject) params.append('subject', subject);
    if (period) params.append('period', period);
    
    return this.request(`/api/v1/ai-advanced/analytics/performance?${params.toString()}`);
  }

  async getEngagementAnalytics(period?: string): Promise<any> {
    const params = new URLSearchParams();
    if (period) params.append('period', period);
    
    return this.request(`/api/v1/ai-advanced/analytics/engagement?${params.toString()}`);
  }

  private async request<T>(endpoint: string, options?: RequestInit): Promise<T> {
    const response = await apiClient.request<T>(endpoint, options);
    return response.data;
  }
}

export const aiAdvancedAPI = new AIAdvancedAPI();

File: backend/api/v1/ai_advanced.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime
from core.database import get_db
from core.security import get_current_user
from models.ai_advanced import (
    AIRecommendation, AITutoringSession, AITutoringInteraction,
    DifficultyDetection, AdaptiveContent
)
from models.analytics import LearningAnalytics
from models.user import User, UserRole
from pydantic import BaseModel

router = APIRouter(tags=["ai_advanced"])

# Pydantic models
class AIRecommendationCreate(BaseModel):
    recommendation_type: str
    title: str
    description: Optional[str] = None
    content_id: Optional[int] = None
    quiz_id: Optional[int] = None
    learning_path_id: Optional[int] = None
    confidence_score: float = 0.0
    reason: Optional[str] = None

class AITutoringSessionCreate(BaseModel):
    subject: Optional[str] = None
    topic: Optional[str] = None
    session_type: str = "general"

class AITutoringInteractionCreate(BaseModel):
    user_message: str
    interaction_type: str = "question"

class DifficultyDetectionCreate(BaseModel):
    subject: str
    topic: str
    difficulty_level: str
    confidence_score: float = 0.0
    evidence: Optional[dict] = None

class AIRecommendationResponse(BaseModel):
    id: int
    user_id: int
    recommendation_type: str
    title: str
    description: Optional[str]
    content_id: Optional[int]
    quiz_id: Optional[int]
    learning_path_id: Optional[int]
    confidence_score: float
    reason: Optional[str]
    is_accepted: bool
    is_dismissed: bool
    created_at: datetime
    
    class Config:
        from_attributes = True

class AITutoringSessionResponse(BaseModel):
    id: int
    user_id: int
    subject: Optional[str]
    topic: Optional[str]
    session_type: str
    start_time: datetime
    end_time: Optional[datetime]
    duration: Optional[int]
    status: str
    notes: Optional[str]
    
    class Config:
        from_attributes = True

class AITutoringInteractionResponse(BaseModel):
    id: int
    session_id: int
    user_message: str
    ai_response: str
    interaction_type: str
    timestamp: datetime
    user_satisfaction: Optional[int]
    
    class Config:
        from_attributes = True

class DifficultyDetectionResponse(BaseModel):
    id: int
    user_id: int
    subject: str
    topic: str
    difficulty_level: str
    confidence_score: float
    evidence: Optional[dict]
    detected_at: datetime
    is_resolved: bool
    resolution_notes: Optional[str]
    
    class Config:
        from_attributes = True

# AI Recommendations
@router.get("/recommendations", response_model=List[AIRecommendationResponse])
async def get_ai_recommendations(
    recommendation_type: Optional[str] = None,
    limit: int = 10,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les recommandations IA pour l'utilisateur"""
    query = db.query(AIRecommendation).filter(
        AIRecommendation.user_id == current_user.id,
        AIRecommendation.is_dismissed == False
    )
    
    if recommendation_type:
        query = query.filter(AIRecommendation.recommendation_type == recommendation_type)
    
    recommendations = query.order_by(AIRecommendation.confidence_score.desc()).limit(limit).all()
    return [AIRecommendationResponse(**rec.__dict__) for rec in recommendations]

@router.post("/recommendations/{recommendation_id}/accept")
async def accept_ai_recommendation(
    recommendation_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Accepter une recommandation IA"""
    recommendation = db.query(AIRecommendation).filter(
        AIRecommendation.id == recommendation_id,
        AIRecommendation.user_id == current_user.id
    ).first()
    
    if not recommendation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Recommandation non trouvée"
        )
    
    recommendation.is_accepted = True
    db.commit()
    
    return {"message": "Recommandation acceptée avec succès"}

@router.post("/recommendations/{recommendation_id}/dismiss")
async def dismiss_ai_recommendation(
    recommendation_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Rejeter une recommandation IA"""
    recommendation = db.query(AIRecommendation).filter(
        AIRecommendation.id == recommendation_id,
        AIRecommendation.user_id == current_user.id
    ).first()
    
    if not recommendation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Recommandation non trouvée"
        )
    
    recommendation.is_dismissed = True
    db.commit()
    
    return {"message": "Recommandation rejetée avec succès"}

# AI Tutoring
@router.post("/tutoring/sessions", response_model=AITutoringSessionResponse)
async def create_ai_tutoring_session(
    session: AITutoringSessionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Créer une session de tutorat IA"""
    db_session = AITutoringSession(
        user_id=current_user.id,
        subject=session.subject,
        topic=session.topic,
        session_type=session.session_type
    )
    
    db.add(db_session)
    db.commit()
    db.refresh(db_session)
    
    return AITutoringSessionResponse(**db_session.__dict__)

@router.get("/tutoring/sessions", response_model=List[AITutoringSessionResponse])
async def get_ai_tutoring_sessions(
    status: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les sessions de tutorat IA"""
    query = db.query(AITutoringSession).filter(AITutoringSession.user_id == current_user.id)
    
    if status:
        query = query.filter(AITutoringSession.status == status)
    
    sessions = query.order_by(AITutoringSession.start_time.desc()).all()
    return [AITutoringSessionResponse(**session.__dict__) for session in sessions]

@router.post("/tutoring/sessions/{session_id}/interactions", response_model=AITutoringInteractionResponse)
async def create_ai_tutoring_interaction(
    session_id: int,
    interaction: AITutoringInteractionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Créer une interaction de tutorat IA"""
    # Vérifier si la session existe et appartient à l'utilisateur
    session = db.query(AITutoringSession).filter(
        AITutoringSession.id == session_id,
        AITutoringSession.user_id == current_user.id
    ).first()
    
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session de tutorat non trouvée"
        )
    
    # Simuler une réponse IA (dans un vrai système, cela appellerait un service IA)
    ai_response = f"Voici une réponse IA à votre question : {interaction.user_message}"
    
    db_interaction = AITutoringInteraction(
        session_id=session_id,
        user_message=interaction.user_message,
        ai_response=ai_response,
        interaction_type=interaction.interaction_type
    )
    
    db.add(db_interaction)
    db.commit()
    db.refresh(db_interaction)
    
    return AITutoringInteractionResponse(**db_interaction.__dict__)

@router.get("/tutoring/sessions/{session_id}/interactions", response_model=List[AITutoringInteractionResponse])
async def get_ai_tutoring_interactions(
    session_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les interactions d'une session de tutorat"""
    # Vérifier si la session existe et appartient à l'utilisateur
    session = db.query(AITutoringSession).filter(
        AITutoringSession.id == session_id,
        AITutoringSession.user_id == current_user.id
    ).first()
    
    if not session:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Session de tutorat non trouvée"
        )
    
    interactions = db.query(AITutoringInteraction).filter(
        AITutoringInteraction.session_id == session_id
    ).order_by(AITutoringInteraction.timestamp).all()
    
    return [AITutoringInteractionResponse(**interaction.__dict__) for interaction in interactions]

# Difficulty Detection
@router.post("/difficulty-detection", response_model=DifficultyDetectionResponse)
async def create_difficulty_detection(
    detection: DifficultyDetectionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Créer une détection de difficulté"""
    db_detection = DifficultyDetection(
        user_id=current_user.id,
        subject=detection.subject,
        topic=detection.topic,
        difficulty_level=detection.difficulty_level,
        confidence_score=detection.confidence_score,
        evidence=detection.evidence
    )
    
    db.add(db_detection)
    db.commit()
    db.refresh(db_detection)
    
    return DifficultyDetectionResponse(**db_detection.__dict__)

@router.get("/difficulty-detection", response_model=List[DifficultyDetectionResponse])
async def get_difficulty_detections(
    subject: Optional[str] = None,
    is_resolved: Optional[bool] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les détections de difficulté"""
    query = db.query(DifficultyDetection).filter(DifficultyDetection.user_id == current_user.id)
    
    if subject:
        query = query.filter(DifficultyDetection.subject == subject)
    if is_resolved is not None:
        query = query.filter(DifficultyDetection.is_resolved == is_resolved)
    
    detections = query.order_by(DifficultyDetection.detected_at.desc()).all()
    return [DifficultyDetectionResponse(**detection.__dict__) for detection in detections]

@router.put("/difficulty-detection/{detection_id}/resolve")
async def resolve_difficulty_detection(
    detection_id: int,
    resolution_notes: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Marquer une détection de difficulté comme résolue"""
    detection = db.query(DifficultyDetection).filter(
        DifficultyDetection.id == detection_id,
        DifficultyDetection.user_id == current_user.id
    ).first()
    
    if not detection:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Détection de difficulté non trouvée"
        )
    
    detection.is_resolved = True
    detection.resolution_notes = resolution_notes
    db.commit()
    
    return {"message": "Détection de difficulté marquée comme résolue"}

# Learning Analytics
@router.get("/analytics/performance")
async def get_performance_analytics(
    subject: Optional[str] = None,
    period: str = "week",
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les analytics de performance"""
    # Ici, vous implémenteriez la logique pour analyser les performances
    # Pour l'instant, retournons des données simulées
    analytics_data = {
        "user_id": current_user.id,
        "period": period,
        "subject": subject,
        "overall_score": 85.5,
        "improvement_rate": 12.3,
        "strengths": ["Mathématiques", "Logique"],
        "weaknesses": ["Histoire", "Géographie"],
        "recommendations": [
            "Continuez à pratiquer les mathématiques",
            "Consacrez plus de temps à l'histoire"
        ]
    }
    
    return analytics_data

@router.get("/analytics/engagement")
async def get_engagement_analytics(
    period: str = "week",
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les analytics d'engagement"""
    # Données simulées d'engagement
    engagement_data = {
        "user_id": current_user.id,
        "period": period,
        "total_time_spent": 1250,  # minutes
        "sessions_count": 15,
        "completion_rate": 78.5,
        "engagement_score": 8.2,
        "trends": {
            "daily_activity": [65, 70, 85, 90, 75, 80, 95],
            "weekly_progress": [12, 15, 18, 22, 25, 28, 30]
        }
    }
    
    return engagement_data 